File: doc/workday/signature-suite-impls/java/signaturesuite/src/main/java/did/Did.java
package did;

import canonical.Canonical;
import net.i2p.crypto.eddsa.EdDSAPublicKey;
import net.i2p.crypto.eddsa.spec.EdDSAPrivateKeySpec;
import net.i2p.crypto.eddsa.spec.EdDSAPublicKeySpec;
import org.bitcoinj.core.Base58;
import proof.Proof;

import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.UUID;

public final class Did extends UnsignedDidDoc {
    public static final String INITIAL_KEY = "key-1";
    public static final String DID_METHOD_PREFIX = "did:work:";

    private Proof proof;

    private Did(final UnsignedDidDoc unsignedDidDoc,
                final Proof proof) {
        super(unsignedDidDoc.getId(), unsignedDidDoc.getPublicKey(), unsignedDidDoc.getAuthentication(), unsignedDidDoc.getService());
        this.proof = proof;
    }

    public static String generateDID(final EdDSAPublicKey pubKey) {
        final byte[] pubKeyBytes = pubKey.getAbyte();
        final byte[] firstSixteen = Arrays.copyOfRange(pubKeyBytes, 0, 16);
        return DID_METHOD_PREFIX + Base58.encode(firstSixteen);
    }

    public static Did signDIDDoc(final UnsignedDidDoc unsignedDidDoc,
                                 final EdDSAPrivateKeySpec privKey,
                                 final String keyRef) throws Exception {
        final String didJson = Canonical.toJson(unsignedDidDoc);
        final String canonicalDidJson = Canonical.canonicalize(didJson);
        final String nonce = UUID.randomUUID().toString();
        final Proof proof = Proof.createEd25519Proof(canonicalDidJson.getBytes(StandardCharsets.UTF_8), keyRef, privKey, nonce);
        return new Did(unsignedDidDoc, proof);
    }

    public static boolean validateDidDocProof(final Did didDoc,
                                              final EdDSAPublicKeySpec pubKey) throws Exception {
        final String didJson = Canonical.toJson(didDoc.getUnsignedDidDoc());
        final String canonicalDidJson = Canonical.canonicalize(didJson);
        return Proof.verifyEd25519Proof(pubKey, didDoc.getProof(), canonicalDidJson.getBytes());
    }

    public UnsignedDidDoc getUnsignedDidDoc() {
        return new UnsignedDidDoc(this.getId(), this.getPublicKey(), this.getAuthentication(), this.getService());
    }

    public Proof getProof() {
        return proof;
    }

    @Override
    public String toString() {
        return Canonical.toJson(this);
    }
}

class UnsignedDidDoc {
    private String id;
    private KeyDef[] publicKey;
    private String[] authentication;
    private ServiceDef[] service;

    public UnsignedDidDoc(final String id,
                          final KeyDef[] publicKey,
                          final String[] authentication,
                          final ServiceDef[] service) {
        this.id = id;
        this.publicKey = publicKey;
        this.authentication = authentication;
        this.service = service;
    }

    public String getId() {
        return id;
    }

    public KeyDef[] getPublicKey() {
        return publicKey;
    }

    public String[] getAuthentication() {
        return authentication;
    }

    public ServiceDef[] getService() {
        return service;
    }

    @Override
    public String toString() {
        return Canonical.toJson(this);
    }
}

class KeyDef {
    private String id;
    private String type;
    private String controller;
    private String publicKeyBase58;

    public KeyDef(final String id,
                  final String type,
                  final String controller,
                  final String publicKeyBase58) {
        this.id = id;
        this.type = type;
        this.controller = controller;
        this.publicKeyBase58 = publicKeyBase58;
    }

    public String getId() {
        return id;
    }

    public String getType() {
        return type;
    }

    public String getController() {
        return controller;
    }

    public String getPublicKeyBase58() {
        return publicKeyBase58;
    }

    @Override
    public String toString() {
        return Canonical.toJson(this);
    }
}

class ServiceDef {
    private String id;
    private String type;
    private String serviceEndpoint;

    public ServiceDef(final String id,
                      final String type,
                      final String serviceEndpoint) {
        this.id = id;
        this.type = type;
        this.serviceEndpoint = serviceEndpoint;
    }

    public String getId() {
        return id;
    }

    public String getType() {
        return type;
    }

    public String getServiceEndpoint() {
        return serviceEndpoint;
    }

    @Override
    public String toString() {
        return Canonical.toJson(this);
    }
}

File: doc/workday/signature-suite-impls/golang/did/did.go
package did

import (
	"github.com/google/uuid"
	"github.com/mr-tron/base58"
	"github.com/pkg/errors"
	"golang.org/x/crypto/ed25519"

	"signaturesuite/canonical"
	"signaturesuite/proof"
)

type UnsignedDIDDoc struct {
	ID             string       `json:"id"`
	PublicKey      []KeyDef     `json:"publicKey"`
	Authentication []string     `json:"authentication"`
	Service        []ServiceDef `json:"service"`
}

// DIDDoc a W3C compliant signed DID Document
type DIDDoc struct {
	*UnsignedDIDDoc
	*proof.Proof `json:"proof"`
}

// KeyDef represents a DID public key
type KeyDef struct {
	ID              string `json:"id"`
	Type            string `json:"type"`
	Controller      string `json:"controller,omitempty"`
	PublicKeyBase58 string `json:"publicKeyBase58"`
}

type ServiceDef struct {
	ID              string `json:"id"`
	Type            string `json:"type"`
	ServiceEndpoint string `json:"serviceEndpoint"`
}

const (
	// InitialKey the key reference assigned to the first key in a DidDoc
	InitialKey      = "key-1"
	IssuerDIDMethod = "did:work:"
)

// GenerateDID generate a DID
func GenerateDID(publicKey *ed25519.PublicKey) string {
	bytes := *publicKey
	return IssuerDIDMethod + base58.Encode(bytes[0:16])
}

func SignDIDDoc(unsignedDoc UnsignedDIDDoc, privKey *ed25519.PrivateKey, keyRef string) (*DIDDoc, error) {
	docBytes, err := canonical.Marshal(&unsignedDoc)
	if err != nil {
		return nil, errors.New("failed to Marshal Unsigned Doc to JSON")
	}
	nonce := uuid.New().String()
	docProof, err := proof.CreateEd25519Proof(docBytes, keyRef, privKey, nonce)
	return &DIDDoc{UnsignedDIDDoc: &unsignedDoc, Proof: docProof}, err
}

func ValidateDIDDocProof(didDoc DIDDoc, pubKey ed25519.PublicKey) error {
	bytes, err := canonical.Marshal(didDoc.UnsignedDIDDoc)
	if err != nil {
		return err
	}
	return proof.VerifyEd25519Proof(pubKey, *didDoc.Proof, bytes)
}
